### ISSUE TITLE
Deploy heads linger around even after app is uninstalled

### ISSUE THREAD

**Issue by uajain** [flatpak/flatpak#3237]
## Linux distribution and version
Fedora 31 and EndlessOS Both
## Flatpak version
Master : 4718bcc
## Description of the problem
`$ref`-named files still linger around in `/var/lib/flatpak/repo/refs/heads/deploy/app` even after the app is uninstalled.

## Steps to reproduce
1) Install an app (`flatpak install flathub org.gnome.Photos`)
2) Verify `ls  -al /var/lib/flatpak/repo/refs/heads/deploy/app/ | grep org.gnome.Photos` shows a `org.gnome.Photos` dir
3) Uninstall (`flatpak uninstall org.gnome.Photos`)
4) Repeat 2), same observations observed.

## Actual results
After un-installing, `$ref`-named files should be cleaned up too.

**Comment by wjt** [comment]
Can you clarify whether you see just the directory (eg) `/var/lib/flatpak/repo/refs/heads/deploy/app/org.gnome.Photos/x86_64`, or whether there is actually a (eg) `stable` file within it?

I can reproduce the tree of empty directories being left around – a bit untidy and would be nice to prune, but not a problem in practice.

**Comment by uajain** [comment]
> Can you clarify whether you see just the directory (eg) `/var/lib/flatpak/repo/refs/heads/deploy/app/org.gnome.Photos/x86_64`, or whether there is actually a (eg) `stable` file within it?
> 
> I can reproduce the tree of empty directories being left around – a bit untidy and would be nice to prune, but not a problem in practice.

Yes, you are right about empty directories. I missed to look what's inside the directory before writing the ticket.

**Comment by mwleeds** [comment]
We also leave behind empty directories when we delete something in `refs/remotes/` or `refs/mirrors/`. It would save a bit of disk space and a bit of performance if we delete those, as I said here: https://github.com/flatpak/flatpak/pull/3225#issuecomment-556824971

**Comment by mwleeds** [comment]
I wonder if the place to implement this would be in libostree, during a prune operation.

**Comment by mwleeds** [comment]
A related issue: #2155

**Comment by alexlarsson** [comment]
I agree with @mwleeds. This should really be done by ostree during pruning.
